modules/firstPartyData/index.js

```diff
@@ -0,0 +1,256 @@
+    let optout = (STORAGE.cookiesAreEnabled() && STORAGE.getCookie('_pubcid_optout')) ||
+      (STORAGE.hasLocalStorage() && STORAGE.getDataFromLocalStorage('_pubcid_optout'));
+
+    if (key === 'user' && optout) {
+      utils.logWarn(`Filtered ${key} data: pubcid optout found`);
+      return result;
+    }
+
+    // Create validated object by looping through ortb2 properties
+    modified = (typeof obj[key] === 'object' && !Array.isArray(obj[key]))
+      ? Object.keys(obj[key]).reduce((combined, keyData) => {
+        let data;
+
+        // If key is user.data, pass into filterData to remove invalid data and return
+        // Else if key is site.content.data, pass into filterData to remove invalid data and return
+        // Else return data unfiltered
+        if (key === 'user' && keyData === 'data') {
+          data = filterData(obj[key][keyData], key);
+
+          if (data) combined[keyData] = data;
+        } else if (key === 'site' && keyData === 'content' && obj[key][keyData].data) {
+          let content = Object.keys(obj[key][keyData]).reduce((merged, contentData) => {
+            if (contentData === 'data') {
+              data = filterData(obj[key][keyData][contentData], key + '.content');
+
+              if (data) merged[contentData] = data;
+            } else {
+              merged[contentData] = obj[key][keyData][contentData];
+            }
+
+            return merged;
+          }, {});
+
+          if (Object.keys(content).length) combined[keyData] = content;
+        } else {
+          combined[keyData] = obj[key][keyData];
+        }
+
+        return combined;
+      }, {}) : obj[key];
+
+    // Check if modified data has data and return
+    if (Object.keys(modified).length) result[key] = modified;
+
+    return result;
+  }, {}));
+
+  // Return validated data
+  return validObject;
+}
+
+/**
+* Resets global ortb2 data
+*/
+export const resetOrtb2 = () => { ortb2 = {} };
+
+/**
+* Sets default values to ortb2 if exists and adds currency and ortb2 setConfig callbacks on init
+*/
+export function init() {
+  setReferer();
+  setPage();
+  setDomain();
+  setDimensions();
+  setKeywords();
+  setCurrency();
+
+  let conf = utils.mergeDeep({}, ortb2, validateFpd(config.getConfig('ortb2')));
+
+  if (!utils.deepEqual(conf, globalConfig)) {
+    config.setConfig({ortb2: conf});
+    globalConfig = {...conf};
+    resetOrtb2();
+  }
+
+  let bidderDuplicate = {...config.getBidderConfig()};
+
+  Object.keys(bidderDuplicate).forEach(bidder => {
+    let modConf = Object.keys(bidderDuplicate[bidder]).reduce((res, key) => {
+      let valid = (key !== 'ortb2') ? bidderDuplicate[bidder][key] : validateFpd(bidderDuplicate[bidder][key]);
+
+      if (valid) res[key] = valid;
+
+      return res;
+    }, {});
+
+    if (Object.keys(modConf).length) config.setBidderConfig({bidders: [bidder], config: modConf});
+  });
+}
+
+function addBidderRequestHook(fn, bidderRequests) {
+  init();
+  fn.call(this, bidderRequests);
+}
```

Comment:
Should there be a check in place to see if publisher has defined `pbjs.setConfig({ ortb2: { ... } })` before we start execution of the init() function?

In some cases, the module may've been included but, the publisher has not defined **ortb2** property, then it doesn't make sense to run these checks.

Reply:
Part of the functionality is validation; however, the other piece here is that default values are set i.e. device data, referer, etc. Meaning that the existence of ortb2 is not required for the module to run.

Reply:
I agree with what you're saying. Earlier I misunderstood this feature to perform checks on ortb2 only if it's present. But regardless of this presence, if this module is present, it'll anyways create a ortb2 object with default values.